File: contrib/tvmop/core/umath.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

# coding: utf-8
import tvm
import topi
from .. import defop, AllTypes, AllTypesButHalf
from .. import assign_by_req, reduce_axes


def compute_exp2(dtype, ndim):
    A = tvm.placeholder([tvm.var() for _ in range(ndim)], name='A', dtype=dtype)
    if dtype in ['float32', 'float64']:
        B = tvm.compute([tvm.var() for _ in range(ndim)],
                        lambda *index: topi.power(2, A[index]), name='B')
    else:
        B = tvm.compute([tvm.var() for _ in range(ndim)],
                        lambda *index: topi.power(2, A[index].astype('float32')).astype(dtype),
                        name='B')
    s = tvm.create_schedule(B.op)
    return s, A, B


@defop(name="exp2", target="cpu", auto_broadcast=True,
       dtype=AllTypes, ndim=list(range(0, 6)))
def exp2(dtype, ndim):
    s, A, B = compute_exp2(dtype, ndim)
    axes = [axis for axis in B.op.axis]
    fused = s[B].fuse(*axes)
    s[B].parallel(fused)
    return s, [A, B]


@defop(name="cuda_exp2", target="cuda", auto_broadcast=True,
       dtype=AllTypesButHalf, ndim=list(range(0, 6)))
def exp2_gpu(dtype, ndim):
    s, A, B= compute_exp2(dtype, ndim)
    s = tvm.create_schedule(B.op)
    axes = [axis for axis in B.op.axis]
    fused = s[B].fuse(*axes)
    bx, tx = s[B].split(fused, factor=64)
    s[B].bind(bx, tvm.thread_axis("blockIdx.x"))
    s[B].bind(tx, tvm.thread_axis("threadIdx.x"))
    return s, [A, B]


def compute_backward_exp2(dtype, ndim):
    A = tvm.placeholder([tvm.var() for _ in range(ndim)], name='A', dtype=dtype)
    B = tvm.placeholder([tvm.var() for _ in range(ndim)], name='B', dtype=dtype)
    C = tvm.placeholder([tvm.var() for _ in range(ndim)], name='C', dtype=dtype)
    D = tvm.compute([tvm.var() for _ in range(ndim)],
                    lambda *index:  A[index] * B[index] * C[index] / 2, name='D')
    s = tvm.create_schedule(D.op)
    return s, A, B, C, D


@defop(name="backward_exp2", target="cpu", auto_broadcast=True,
       dtype=AllTypes, ndim=list(range(0, 6)))
def backward_exp2(dtype, ndim):
    s, A, B, C, D = compute_backward_exp2(dtype, ndim)
    axes = [axis for axis in D.op.axis]
    fused = s[D].fuse(*axes)
    s[D].parallel(fused)
    return s, [A, B, C, D]


@defop(name="cuda_backward_exp2", target="cuda", auto_broadcast=True,
       dtype=AllTypesButHalf, ndim=list(range(0, 6)))
def backward_exp2_gpu(dtype, ndim):
    s, A, B, C, D= compute_backward_exp2(dtype, ndim)
    s = tvm.create_schedule(D.op)
    axes = [axis for axis in D.op.axis]
    fused = s[D].fuse(*axes)
    bx, tx = s[D].split(fused, factor=64)
    s[D].bind(bx, tvm.thread_axis("blockIdx.x"))
    s[D].bind(tx, tvm.thread_axis("threadIdx.x"))
    return s, [A, B, C, D]

File: src/operator/contrib/tvmop/core/umath.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * Copyright (c) 2019 by Contributors
 * \file umath.cc
 * \brief
 * \author Guangyong Shi
 */
#ifdef MXNET_USE_TVM_OP
#include <tvm/runtime/packed_func.h>
#include <tvm/runtime/registry.h>
#include <tvm/runtime/c_runtime_api.h>
#include <mxnet/base.h>
#include <string>
#include "../../../tensor/elemwise_binary_broadcast_op.h"
#include "../../../tvmop/op_module.h"
#include "../../../tensor/elemwise_binary_op.h"

namespace mxnet {
namespace op {

static constexpr char func_exp2_cpu[] = "exp2";
static constexpr char func_epx2_gpu[] = "cuda_exp2";
static constexpr char func_backward_epx2_cpu[] = "backward_exp2";
static constexpr char func_backward_epx2_gpu[] = "cuda_backward_exp2";

template<const char* func>
void TVMOpExp2Compute(const nnvm::NodeAttrs& attrs,
                      const mxnet::OpContext& ctx,
                      const std::vector<TBlob>& inputs,
                      const std::vector<OpReqType>& req,
                      const std::vector<TBlob>& outputs) {
  CHECK_EQ(inputs.size(), 1U);
  CHECK_EQ(outputs.size(), 1U);
  if (outputs[0].Size() == 0U) return;
  tvm::runtime::TVMOpModule::Get()->Call(func, ctx, {inputs[0], outputs[0]});
}

template<const char* func>
void TVMExp2Backward(const nnvm::NodeAttrs& attrs,
                     const OpContext& ctx,
                     const std::vector<TBlob>& inputs,
                     const std::vector<OpReqType>& req,
                     const std::vector<TBlob>& outputs) {
  CHECK_EQ(inputs.size(), 3U);
  CHECK_EQ(outputs.size(), 1U);
  CHECK_EQ(req.size(), 1U);
  if (inputs[0].Size() == 0U) return;
  using namespace mshadow;
  const TBlob& out_grad = inputs[0];
  const TBlob& in_data1 = inputs[1];
  const TBlob& in_data2 = inputs[2];
  const TBlob& in_grad = outputs[0];
  tvm::runtime::TVMOpModule::Get()->Call(func, ctx, {out_grad, in_data1, in_data2, in_grad});
}

NNVM_REGISTER_OP(_npi_exp2)
    .set_num_inputs(1)
    .set_num_outputs(1)
    .add_argument("data", "NDArray-or-Symbol", "resources")
    .set_attr<mxnet::FInferShape>("FInferShape", mxnet::op::ElemwiseShape<1, 1>)
    .set_attr<nnvm::FInferType>("FInferType", mxnet::op::ElemwiseType<1, 1>)
#if MXNET_USE_CUDA
    .set_attr<mxnet::FCompute>("FCompute<gpu>", mxnet::op::TVMOpExp2Compute<func_epx2_gpu>)
#endif  // MXNET_USE_CUDA
    .set_attr<mxnet::FCompute>("FCompute<cpu>", mxnet::op::TVMOpExp2Compute<func_exp2_cpu>)
    .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseInOut{ "_backward_exp2" });

NNVM_REGISTER_OP(_backward_exp2)
    .set_num_inputs(3)
    .set_num_outputs(1)
    .set_attr<nnvm::TIsBackward>("TIsBackward", true)
#if MXNET_USE_CUDA
    .set_attr<FCompute>("FCompute<gpu>", mxnet::op::TVMExp2Backward<func_backward_epx2_gpu>)
#endif  // MXNET_USE_CUDA
    .set_attr<FCompute>("FCompute<cpu>", mxnet::op::TVMExp2Backward<func_backward_epx2_cpu>);

}  // namespace op
}  // namespace mxnet

#endif  // MXNET_USE_TVM_OP
